### PR TITLE
fix(bootable): don't generate lazy content until it's rendered

### LINKED ISSUE #8712: [Feature Request] [VDialog] [Performance] Don't create ThemeProvider when dialog is lazy and not visible
### Problem to solve
VDialog components are creating a ThemeProvider even when the dialog is hidden. 

This is an extra component allocation, and it also causes extra dependency subscriptions against `$vuetify` theme props. On a page that has a large number of lazy dialogs (in my case, a table with multiple button-activated dialogs on each row), this causes a significant performance hit due to https://github.com/vuejs/vue/issues/10435

I can work around this by taking my dialogs out of my table and manually providing data and activating, but it would be nice if this wasn't needed.

### Proposed solution
Include the ThemeProvider (and maybe even its containing div(s)) in the set of elements that aren't rendered when the dialog is hidden.

### PR DESCRIPTION
**Pull request by KaelWD**
## Motivation and Context

fixes #8712

better performance, fewer empty DOM elements, skips a bunch of calculations for hidden elements

## Markup:

<details>

```vue
<template>
  <v-container>
    <v-dialog
      v-model="dialog"
      width="500"
    >
      <template v-slot:activator="{ on }">
        <v-btn
          color="red lighten-2"
          dark
          v-on="on"
        >
          Dialog
        </v-btn>
      </template>

      <v-card>
        <v-card-title
          class="headline grey lighten-2"
          primary-title
        >
          Privacy Policy
        </v-card-title>

        <v-card-text>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </v-card-text>

        <v-divider></v-divider>

        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn
            color="primary"
            text
            @click="dialog = false"
          >
            I accept
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>

    <v-tooltip bottom>
      <template v-slot:activator="{ on }">
        <v-btn color="primary" dark v-on="on">Tooltip</v-btn>
      </template>
      <span>Tooltip</span>
    </v-tooltip>

    <v-menu offset-y>
      <template v-slot:activator="{ on }">
        <v-btn
          color="primary"
          dark
          v-on="on"
        >
          Menu
        </v-btn>
      </template>
      <v-list>
        <v-list-item
          v-for="(item, index) in items"
          :key="index"
          @click=""
        >
          <v-list-item-title>{{ item.title }}</v-list-item-title>
        </v-list-item>
      </v-list>
    </v-menu>

    <v-expansion-panels>
      <v-expansion-panel
        v-for="(item,i) in 5"
        :key="i"
      >
        <v-expansion-panel-header>Item</v-expansion-panel-header>
        <v-expansion-panel-content>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
        </v-expansion-panel-content>
      </v-expansion-panel>
    </v-expansion-panels>

    <v-card
      max-width="500"
      class="mx-auto"
    >
      <v-toolbar
        color="teal"
        dark
      >
        <v-app-bar-nav-icon></v-app-bar-nav-icon>

        <v-toolbar-title>Topics</v-toolbar-title>

        <v-spacer></v-spacer>

        <v-btn icon>
          <v-icon>mdi-dots-vertical</v-icon>
        </v-btn>
      </v-toolbar>

      <v-list>
        <v-list-group
          v-for="item in items"
          :key="item.title"
          v-model="item.active"
          :prepend-icon="item.action"
          no-action
        >
          <template v-slot:activator>
            <v-list-item-content>
              <v-list-item-title v-text="item.title"></v-list-item-title>
            </v-list-item-content>
          </template>

          <v-list-item
            v-for="subItem in item.items"
            :key="subItem.title"
            @click=""
          >
            <v-list-item-content>
              <v-list-item-title v-text="subItem.title"></v-list-item-title>
            </v-list-item-content>
          </v-list-item>
        </v-list-group>
      </v-list>
    </v-card>

    <v-carousel
      cycle
      height="400"
      hide-delimiter-background
      show-arrows-on-hover
    >
      <v-carousel-item
        v-for="(slide, i) in slides"
        :key="i"
      >
        <v-sheet
          :color="colors[i]"
          height="100%"
        >
          <v-row
            class="fill-height"
            align="center"
            justify="center"
          >
            <div class="display-3">{{ slide }} Slide</div>
          </v-row>
        </v-sheet>
      </v-carousel-item>
    </v-carousel>

    <v-card flat tile>
      <v-window v-model="onboarding" vertical>
        <v-window-item
          v-for="n in length"
          :key="`card-${n}`"
        >
          <v-card
            color="grey"
            height="200"
          >
            <v-row
              class="fill-height"
              align="center"
              justify="center"
              tag="v-card-text"
            >
              <h1 style="font-size: 5rem;" class="white--text">Slide {{ n }}</h1>
            </v-row>
          </v-card>
        </v-window-item>
      </v-window>

      <v-card-actions class="justify-space-between">
        <v-btn
          text
          @click="prev"
        >
          <v-icon>mdi-chevron-left</v-icon>
        </v-btn>
        <v-item-group
          v-model="onboarding"
          class="text-center"
          mandatory
        >
          <v-item
            v-for="n in length"
            :key="`btn-${n}`"
            v-slot:default="{ active, toggle }"
          >
            <v-btn
              :input-value="active"
              icon
              @click="toggle"
            >
              <v-icon>mdi-record</v-icon>
            </v-btn>
          </v-item>
        </v-item-group>
        <v-btn
          text
          @click="next"
        >
          <v-icon>mdi-chevron-right</v-icon>
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-container>
</template>

<script>
  export default {
    data: () => ({
      dialog: false,

      items: [
        {
          action: 'local_activity',
          title: 'Attractions',
          items: [
            { title: 'List Item' },
          ],
        },
        {
          action: 'restaurant',
          title: 'Dining',
          active: true,
          items: [
            { title: 'Breakfast & brunch' },
            { title: 'New American' },
            { title: 'Sushi' },
          ],
        },
        {
          action: 'school',
          title: 'Education',
          items: [
            { title: 'List Item' },
          ],
        },
        {
          action: 'directions_run',
          title: 'Family',
          items: [
            { title: 'List Item' },
          ],
        },
        {
          action: 'healing',
          title: 'Health',
          items: [
            { title: 'List Item' },
          ],
        },
        {
          action: 'content_cut',
          title: 'Office',
          items: [
            { title: 'List Item' },
          ],
        },
        {
          action: 'local_offer',
          title: 'Promotions',
          items: [
            { title: 'List Item' },
          ],
        },
      ],

      colors: [
        'indigo',
        'warning',
        'pink darken-2',
        'red lighten-1',
        'deep-purple accent-4',
      ],
      slides: [
        'First',
        'Second',
        'Third',
        'Fourth',
        'Fifth',
      ],

      length: 3,
      onboarding: 0,
    }),
    methods: {
      next () {
        this.onboarding = this.onboarding + 1 === this.length
          ? 0
          : this.onboarding + 1
      },
      prev () {
        this.onboarding = this.onboarding - 1 < 0
          ? this.length - 1
          : this.onboarding - 1
      },
    },
  }
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

### CODE REVIEW

**Comment by ascott18** [comment]
Wouldn't it be best to use a direct `if (this.hasContent) { ... }` check instead of `showLazyContent`? Because unless I'm mistaken, with `showLazyContent` the whole VNode tree still gets created in this case (but just not patched into the DOM because `showLazyContent` will just throw it away.)

**Comment by KaelWD** [comment]
Ah right good catch

**Comment by ascott18** [comment]
Would love to see this get wrapped up if you guys can find the time! Especially the change to replace the usage `showLazyContent` with direct branching (or change `showLazyContent` to accept a function that will lazily create the VNode tree only if needed).

I understand that you guys are busy with all sorts of other good stuff, though, and I appreciate all the work that you do!

**Comment by johnleider** [comment]
Pulse check.

**Comment by johnleider** [comment]
This Pull Request is being closed due to inactivity.

**Comment by KaelWD** [comment]
Related: #10201

File: packages/vuetify/src/components/VCarousel/__tests__/__snapshots__/VCarouselItem.spec.ts.snap [hunk 1]
```diff
@@ -1,15 +1,5 @@
...
-`;
+exports[`VCarouselItem.ts should render component and match snapshot 1`] = `undefined`;
 
-exports[`VCarouselItem.ts should throw warning when not used inside v-carousel 1`] = `
-<div class="v-window-item"
-     style="display: none;"
->
-</div>
-`;
+exports[`VCarouselItem.ts should throw warning when not used inside v-carousel 1`] = `undefined`;
```
Comment (sh7dm) [inline packages/vuetify/src/components/VCarousel/__tests__/__snapshots__/VCarouselItem.spec.ts.snap]: doesn't look correct. If you check for `undefined`, use a matcher for that

Comment (KaelWD) [reply packages/vuetify/src/components/VCarousel/__tests__/__snapshots__/VCarouselItem.spec.ts.snap]: snapshots can all go to hell

File: packages/vuetify/src/components/VDialog/VDialog.ts [hunk 2]
```diff
@@ -241,68 +241,55 @@ export default baseMixins.extend({
...
-      class: this.classes,
-      ref: 'dialog',
-      directives: [
-        {
-          name: 'click-outside',
-          value: this.onClickOutside,
-          args: {
-            closeConditional: this.closeConditional,
-            include: this.getOpenDependentElements,
+    let dialog = this.showLazyContent(() => {
```
Comment (johnleider) [inline packages/vuetify/src/components/VDialog/VDialog.ts]: I feel like this should live in its own method. it makes the render fn feel overly verbose.

File: packages/vuetify/src/components/VDialog/VDialog.ts [hunk 3]
```diff
@@ -241,68 +241,55 @@ export default baseMixins.extend({
...
-    children.push(this.genActivator())
+      if (!this.fullscreen) {
+        data.style = {
+          maxWidth: this.maxWidth === 'none' ? undefined : convertToUnit(this.maxWidth),
+          width: this.width === 'auto' ? undefined : convertToUnit(this.width),
+        }
+      }
 
-    let dialog = h('div', data, this.showLazyContent(this.getContentSlot()))
+      return [
```
Comment (johnleider) [inline packages/vuetify/src/components/VDialog/VDialog.ts]: this should be in its own method as well

File: packages/vuetify/src/components/VDialog/VDialog.ts [hunk 4]
```diff
@@ -312,6 +299,21 @@ export default baseMixins.extend({
           this.attach === 'attach',
       },
       attrs: { role: 'dialog' },
-    }, children)
+    }, [
+      this.genActivator(),
+      h('div', {
```
Comment (johnleider) [inline packages/vuetify/src/components/VDialog/VDialog.ts]: we already have `genActivator`, this should be in its own method as well

File: packages/vuetify/src/components/VMenu/VMenu.ts [hunk 5]
```diff
@@ -350,7 +350,13 @@ export default baseMixins.extend({
       return this.$createElement(
         'div',
         options,
-        this.showLazyContent(this.getContentSlot())
+        [this.$createElement(VThemeProvider, {
+          props: {
+            root: true,
```
Comment (johnleider) [inline packages/vuetify/src/components/VMenu/VMenu.ts]: this should be alpha sorted

File: packages/vuetify/src/components/VSelect/__tests__/VSelect.spec.ts [hunk 6]
```diff
@@ -268,7 +268,7 @@ describe('VSelect.ts', () => {
 
     const wrapper = mountFunction({
       propsData: {
-        menuProps: { contentClass: 'v-menu-class' },
+        menuProps: { contentClass: 'v-menu-class', eager: true },
```
Comment (johnleider) [inline packages/vuetify/src/components/VSelect/__tests__/VSelect.spec.ts]: this should be on 2 lines

File: packages/vuetify/src/components/VTooltip/VTooltip.ts [hunk 7]
```diff
@@ -185,21 +185,23 @@ export default mixins(Colorable, Delayable, Dependent, Detachable, Menuable, Tog
...
-      },
-      style: this.styles,
-      attrs: this.getScopeIdAttrs(),
-      directives: [{
-        name: 'show',
-        value: this.isContentActive,
-      }],
-      ref: 'content',
-    }), this.showLazyContent(this.getContentSlot()))
+    const tooltip = this.showLazyContent(() => [
```
Comment (johnleider) [inline packages/vuetify/src/components/VTooltip/VTooltip.ts]: this should be in its own method

File: packages/vuetify/src/components/VList/VListGroup.ts [hunk 8]
```diff
@@ -160,16 +161,16 @@ export default baseMixins.extend<options>().extend({
...
-      return this.$createElement('div', {
-        staticClass: 'v-list-group__items',
-        directives: [{
-          name: 'show',
-          value: this.isActive,
-        }],
-      }, this.showLazyContent([
-        this.$createElement('div', this.$slots.default),
-      ]))
+    genItems (): VNode[] {
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/VListGroup.ts]:
```suggestion
    genItems (): VNode[] | undefined {
```

If you make this `VNode[] | undefined`, you can remove the need to invoke `this.$createElement()` here: https://github.com/vuetifyjs/vuetify/pull/8823/files#diff-176112ec6d020b2c188161c378061d85R51

File: packages/vuetify/src/mixins/bootable/index.ts [hunk 9]
```diff
@@ -47,8 +47,8 @@ export default Vue.extend<Vue & Toggleable>().extend({
   },
 
   methods: {
-    showLazyContent (content?: VNode[]): VNode[] | undefined {
-      return this.hasContent ? content : undefined
+    showLazyContent (content?: () => VNode[]): VNode[] {
+      return (this.hasContent && content) ? content() : [this.$createElement()]
```
Comment (johnleider) [inline packages/vuetify/src/mixins/bootable/index.ts]:
```suggestion
      return (this.hasContent && content) ? content() : undefined
```

Following the comment made here: https://github.com/vuetifyjs/vuetify/pull/8823/files#diff-a5bc4eebd33025bd247fef3ba12f27dbR164

File: packages/vuetify/src/mixins/bootable/index.ts [hunk 10]
```diff
@@ -47,8 +47,8 @@ export default Vue.extend<Vue & Toggleable>().extend({
   },
 
   methods: {
-    showLazyContent (content?: VNode[]): VNode[] | undefined {
-      return this.hasContent ? content : undefined
+    showLazyContent (content?: () => VNode[]): VNode[] {
```
Comment (johnleider) [inline packages/vuetify/src/mixins/bootable/index.ts]:
```suggestion
    showLazyContent (content?: () => VNode[]): VNode[] | undefined {
```

File: packages/vuetify/src/components/VSelect/__tests__/VSelect.spec.ts (hunk 6 above)
Comment (johnleider) [reply packages/vuetify/src/components/VSelect/__tests__/VSelect.spec.ts]: Disregard

File: packages/vuetify/src/components/VList/VListGroup.ts (hunk 8 above)
Comment (KaelWD) [reply packages/vuetify/src/components/VList/VListGroup.ts]: I was copying Vue, v-if renders a comment when false so it can be reused.

Comment (johnleider) [reply packages/vuetify/src/components/VList/VListGroup.ts]: 👍

File: packages/vuetify/src/components/VMenu/VMenu.ts (hunk 5 above)
Comment (KaelWD) [reply packages/vuetify/src/components/VMenu/VMenu.ts]: nah

File: packages/vuetify/src/components/VDialog/VDialog.ts (hunk 2 above)
Comment (KaelWD) [reply packages/vuetify/src/components/VDialog/VDialog.ts]: I wasn't trying to do a full rewrite or anything, these functions hadn't really changed at all.

**Pull request merged**